### PR TITLE
Add InnerBlocks locking

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
The feature is required by #6993.
This PR adds the possibility to set locking in InnerBlocks in a similar way we can now do with the template locking.
Usage:

```js
edit() {
	return el( 'div', { style: { outline: '1px solid gray', padding: 5 } },
		el( InnerBlocks, {
			locking: 'insert', // all, insert, or none
			template: [
				[ 'core/paragraph', {
					placeholder: 'Enter locking insert p1 content…',
				} ],
				[ 'core/paragraph', {
					placeholder: 'Enter locking insert p2 content…',
				} ],
			],
		} )
	);
}
```

Locking can be all, insert or not set.
Right now, with template locking if we have locking active and a template with a column block we cannot insert anything into the column so it seems locking applies to children.
We followed the same logic in this PR locking applies to descends even if the descends set a less restrictive locking or no locking at all. We may change this behavior but if changing it we should also change it for the CPT templates case.

## How has this been tested?

The best way is to use the test blocks in https://gist.github.com/jorgefilipecosta/2bac096fe3be7f6ff0de7452292161cb, and verify everything works as expected. To simplest way to have the blocks is to paste all the code in the developer console.

### CODE REVIEW

File: editor/utils/block-list.js [hunk 1]
```diff
@@ -40,10 +59,19 @@ const INNER_BLOCK_LIST_CACHE = {};
...
-				} );
+			constructor( props ) {
+				super( props );
+				this.state = {};
+			}
+
+			static getDerivedStateFromProps( props ) {
+				const newSettings = {
+					allowedBlocks: props.allowedBlocks,
+					locking: getHighestLocking( props.locking, props.parentLocking ),
```
Comment (youknowriad) [inline editor/utils/block-list.js]: Why do we compare the `locking` prop with the parent `locking` prop? Shouldn't we just always use the prop passed to `InnerBlocks`?

File: editor/store/selectors.js [hunk 2]
```diff
@@ -1230,20 +1230,20 @@ export const canInsertBlockType = createSelector(
...
-		const isEditorLocked = !! templateLock;
-		if ( isEditorLocked ) {
+		const isLocked = !! getLocking( state, parentUID );
+		if ( isLocked ) {
 			return false;
 		}
 
 		const parentBlockListSettings = getBlockListSettings( state, parentUID );
-		const parentAllowedBlocks = get( parentBlockListSettings, [ 'supportedBlocks' ] );
+		const parentAllowedBlocks = get( parentBlockListSettings, [ 'allowedBlocks' ] );
```
Comment (youknowriad) [inline editor/store/selectors.js]: Is this like an unrelated fix?

Comment (youknowriad) [reply editor/store/selectors.js]: Oh I see, it's renamed nvm

File: editor/store/selectors.js [hunk 3]
```diff
@@ -1744,3 +1744,21 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
 export function getEditorSettings( state ) {
 	return state.settings;
 }
+
+/*
+ * Returns the editor settings.
```
Comment (tofumatt) [inline editor/store/selectors.js]: This comment seems inaccurate; maybe it was a copy/paste oversight?

This seems to just return the locking setting for this block.

**Comment by youknowriad** [comment]
I'm glad to see this being worked on. I wonder if we should support setting this prop in the "template" as well? Like we currently do for the children blocks.

File: editor/store/selectors.js [hunk 4]
```diff
@@ -1744,3 +1744,21 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
...
+ *
+ * @param {Object} state Editor state.
+ *
+ * @return {Object} The editor settings object
```
Comment (tofumatt) [inline editor/store/selectors.js]: This appears to return the locking setting, which is a string (with particular values). Looks like this needs updating.

File: editor/store/selectors.js [hunk 5]
```diff
@@ -1744,3 +1744,21 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
 export function getEditorSettings( state ) {
 	return state.settings;
 }
+
+/*
+ * Returns the editor settings.
+ *
+ * @param {Object} state Editor state.
```
Comment (tofumatt) [inline editor/store/selectors.js]: This takes two arguments but only one is listed.

File: editor/store/selectors.js [hunk 6]
```diff
@@ -1744,3 +1744,21 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
...
+ *
+ * @return {Object} The editor settings object
+ */
+export function getLocking( state, rootUID ) {
```
Comment (tofumatt) [inline editor/store/selectors.js]: I see this function being called using `getLocking( rootUID )` a few times. Am I misunderstanding something?

File: editor/components/block-drop-zone/index.js [hunk 7]
```diff
@@ -137,11 +137,10 @@ export default compose(
...
 		};
 	} ),
-	withSelect( ( select ) => {
-		const { templateLock } = select( 'core/editor' ).getEditorSettings();
-
+	withSelect( ( select, { rootUID } ) => {
+		const { getLocking } = select( 'core/editor' );
 		return {
-			isLocked: !! templateLock,
+			isLocked: !! getLocking( rootUID ),
```
Comment (tofumatt) [inline editor/components/block-drop-zone/index.js]: The function signature for `getLocking` looks like `getLocking( state, rootUID )`… seems like this is wrong.

File: editor/store/selectors.js [hunk 8]
```diff
@@ -1744,3 +1744,21 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
...
+	if ( ! rootUID ) {
+		return getTemplateLock( state );
+	}
+	const blockListSettings = getBlockListSettings( state, rootUID );
+	if ( ! blockListSettings ) {
+		return;
```
Comment (tofumatt) [inline editor/store/selectors.js]: Why return `void`? Wouldn't it make sense to at least return `null`? It seems more semantically valid...

**Review by tofumatt — DISMISSED** [review]
Maybe I'm missing something but the documentation and usage of the `getLocking` function seems very inconsistent. I'm not quite following it.

File: editor/store/selectors.js (hunk 6 above)
Comment (youknowriad) [reply editor/store/selectors.js]: The state is automatically injected by the data module

Comment (youknowriad) [reply editor/store/selectors.js]: (see the README here https://github.com/WordPress/gutenberg/tree/master/data)

Comment (tofumatt) [reply editor/store/selectors.js]: Oh, okay. That is really hard to grok from reading the code 😢

**Comment by tofumatt** [comment]
Looks like the docs for `getLocking` are still off, but I see now I didn't understand how the data module worked.

File: editor/utils/block-list.js (hunk 1 above)
Comment (jorgefilipecosta) [reply editor/utils/block-list.js]: Hi @youknowriad, this is what allows us to propagate the locking to descendants. Even if a child block sets no locking at all or a less restrictive locking the locking of the parent applies. So for example, if the parent adds locking and uses columns block the locking applies also to columns. I'm not certain if this is the right behavior and in fact, I'm now considering changing to just propagate the locking if the child block did not set locking at all.

File: editor/store/selectors.js (hunk 3 above)
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: True I totally missed the update to the comment. Thank you for the catch.

File: editor/utils/block-list.js (hunk 1 above)
Comment (youknowriad) [reply editor/utils/block-list.js]: yes, I'd consider `props.locking || props.parentLocking` a better approach here :) (or something in that vein. We could also question whether propagation is something we want (I know we do this at the moment)

File: editor/store/selectors.js [hunk 9]
```diff
@@ -1744,3 +1744,22 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
...
+
+/*
+ * Returns the locking in the context of a given root block.
+ *
+ * @param {Object}  state   Editor state.
+ * @param {?string} rootUID Block UID.
+ *
+ * @return {?string} Locking in the context of a given block.
+ */
+export function getLocking( state, rootUID ) {
```
[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: The locking states can be all, insert or undefined/none.

[the remark this answers is not in the document]
Comment (mtias) [reply editor/store/selectors.js]: That would be more clear, yes: `getLockedState`. The other concern is that it seems slightly backwards to consider the locked state as the property when most blocks are going to be unlocked.

File: editor/components/block-list/insertion-point.js [hunk 10]
```diff
@@ -92,13 +92,13 @@ export default compose(
 		);
 
 		return {
-			templateLock: getEditorSettings().templateLock,
+			locking: getLockedState( insertionPoint.rootUID ),
```
Comment (mcsf) [inline editor/components/block-list/insertion-point.js]: Prop `locking` should also have its name updated in accordance with `getLockedState`, _or_ should be renamed entirely to something like `isLocked` with `isLocked: !! getLockedState( … )`

File: editor/components/inner-blocks/index.js [hunk 11]
```diff
@@ -3,8 +3,8 @@
  */
 import { withContext } from '@wordpress/components';
 
-function InnerBlocks( { BlockList, layouts, allowedBlocks, template } ) {
-	return <BlockList { ...{ layouts, allowedBlocks, template } } />;
+function InnerBlocks( { BlockList, layouts, allowedBlocks, locking, template } ) {
+	return <BlockList { ...{ layouts, allowedBlocks, locking, template } } />;
```
Comment (mcsf) [inline editor/components/inner-blocks/index.js]: See previous comments on naming.

File: editor/store/selectors.js [hunk 12]
```diff
@@ -1744,3 +1744,22 @@ export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
...
+/*
+ * Returns the locked state in the context of a given root block.
+ *
+ * @param {Object}  state   Editor state.
+ * @param {?string} rootUID Block UID.
+ *
+ * @return {?string} Locked state in the context of a given block.
+ */
+export function getLockedState( state, rootUID ) {
+	if ( ! rootUID ) {
```
Comment (mcsf) [inline editor/store/selectors.js]: Should we make `rootUID` an explicitly optional param so we can avoid the check for falsy values?

```js
getLockedState( state, rootUID = false ) {
  if ( rootUID === false ) {
```

File: editor/utils/block-list.js [hunk 13]
```diff
@@ -98,10 +99,11 @@ export function createInnerBlockList( uid, renderBlockMenu = noop ) {
 
 		const InnerBlockListComponentContainer = compose(
 			withSelect( ( select ) => {
-				const { getBlock, getBlockListSettings } = select( 'core/editor' );
+				const { getBlock, getBlockListSettings, getBlockRootUID, getLockedState } = select( 'core/editor' );
 				return {
 					block: getBlock( uid ),
 					blockListSettings: getBlockListSettings( uid ),
+					parentLocking: getLockedState( getBlockRootUID( uid ) ),
```
Comment (mcsf) [inline editor/utils/block-list.js]: See previous comments on naming.

File: editor/utils/block-list.js [hunk 14]
```diff
@@ -40,10 +41,19 @@ const INNER_BLOCK_LIST_CACHE = {};
...
+			}
+
+			static getDerivedStateFromProps( props ) {
+				const newSettings = {
+					allowedBlocks: props.allowedBlocks,
+					locking: props.locking === undefined ? props.parentLocking : props.locking,
+				};
+				if ( ! isShallowEqual( props.blockListSettings, newSettings ) ) {
+					props.updateNestedSettings( newSettings );
+				}
```
Comment (mcsf) [inline editor/utils/block-list.js]: On usage of `getDerivedStateFromProps`:
- [Strictly speaking](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops), it should return `null` instead of `undefined`.
- Side effects [are to be avoided](https://reactjs.org/blog/2018/05/23/react-v-16-4.html#1-avoid-side-effects-in-getderivedstatefromprops), and that is exactly what this particular method is doing. :)

File: editor/components/block-list/insertion-point.js (hunk 10 above)
Comment (jorgefilipecosta) [reply editor/components/block-list/insertion-point.js]: Naming was updated.

File: editor/components/inner-blocks/index.js (hunk 11 above)
Comment (jorgefilipecosta) [reply editor/components/inner-blocks/index.js]: I changed the prop to "lock", in templates we use template_lock so I think using "lock" here is consistent. Unfortunately using isLocked in this case is not possible because we are not setting a boolean value.

File: editor/store/selectors.js (hunk 12 above)
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: Hi @mcsf,
I followed your suggestion with a small change:
```
export function getLockedState( state, rootUID = null ) {
	if ( rootUID === null ) {
```

Sometimes we use other selectors to know the rootUID and they may return null. In that case, we should consider that no rootUID was passed. Using this approach we can avoid extra checks before calling this selector (making the code simpler).

File: editor/utils/block-list.js (hunk 14 above)
Comment (jorgefilipecosta) [reply editor/utils/block-list.js]: Hi @mcsf,
You are totally right it looks like getDerivedStateFromProps should be a pure function without side effects.
I updated the code to use componentDidUpdate as going back to componentWillReceive props is not a long-term option. componentDidUpdate also does not seems like a good way as we may have double render executions but it seems like the best option we have and according to the react docs users don't see the intermediate dom state.

File: editor/components/block-list/insertion-point.js (hunk 10 above)
Comment (aduth) [reply editor/components/block-list/insertion-point.js]: - At the editor level, we name template locking `templateLock`
- At the block level, we name template locking `lock`

File: editor/components/inner-blocks/index.js (hunk 11 above)
Comment (aduth) [reply editor/components/inner-blocks/index.js]: >I changed the prop to "lock", in templates we use template_lock, so I think using "lock" here is consistent. Unfortunately using isLocked in this case is not possible because we are not setting a boolean value.

But in the context of InnerBlocks, isn't the locking still referring to the template?

**Comment by aduth** [comment]
Would it be reasonable to represent inner block locking by `allowedBlocks={ [] }` ?

**Comment by jorgefilipecosta** [comment]
> Would it be reasonable to represent inner block locking by allowedBlocks={ [] } ?

There are two types of locking "all" and "insert" and this syntax does not allow to represent both types. Besides that having allowedBlocks={ [] } does not in fact prevent from inserting other blocks inside, because according to our API allowedBlocks does not prevent children from being inserted. In fact I expect many blocks setting allowedBlocks={ [] } while having many child blocks that we can insert inside.

File: editor/store/selectors.js (hunk 12 above)
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: I had to revert this change and use the false check because I noticed a regression. Some selectors may return an empty string if rootUID is not set e.g. getBlockRootUID( uid ) returns an empty string for top-level blocks.
The best/simplest way to check for undefined, null or empty string seems to be a false check.

**Comment by jorgefilipecosta** [comment]
This PR was rebased and most of the feedback are addressed. It is ready for a new look :)

File: editor/components/inner-blocks/index.js [hunk 15]
```diff
@@ -43,15 +43,19 @@ class InnerBlocks extends Component {
 	updateNestedSettings() {
 		const {
 			blockListSettings,
-			allowedBlocks: nextAllowedBlocks,
+			allowedBlocks,
+			lock,
```
Comment (aduth) [inline editor/components/inner-blocks/index.js]: Mentioned at https://github.com/WordPress/gutenberg/pull/6991#discussion_r196916721: If we call this `templateLock` at the editor level, why do we not call it the same here?

File: editor/components/inner-blocks/index.js [hunk 16]
```diff
@@ -43,15 +43,19 @@ class InnerBlocks extends Component {
...
+			parentLock,
```
Comment (aduth) [inline editor/components/inner-blocks/index.js]: Where is `parentLock` being passed?

Comment (jorgefilipecosta) [reply editor/components/inner-blocks/index.js]: Sorry I missed the commit of parentLock retrieval during the rebase. It was corrected.

File: editor/components/inner-blocks/index.js (hunk 15 above)
Comment (jorgefilipecosta) [reply editor/components/inner-blocks/index.js]: My idea was to try to move locking out of being a template concept. I wanted the lock to be a setting blocks can make use. E.g. have settings that enable or disable locking.
But yes given that now it's main use case is templates I updated the name to be templateLock.

File: editor/store/selectors.js [hunk 17]
```diff
@@ -1192,12 +1192,22 @@ export function getTemplate( state ) {
...
+ *
  * @return {?string}        Block Template Lock
  */
-export function getTemplateLock( state ) {
-	return state.settings.templateLock;
+export function getTemplateLock( state, rootUID ) {
+	if ( ! rootUID ) {
+		return state.settings.templateLock;
+	}
+	const blockListSettings = getBlockListSettings( state, rootUID );
```
Comment (aduth) [inline editor/store/selectors.js]: There aren't any unit tests for this new behavior.

File: editor/components/inner-blocks/index.js [hunk 18]
```diff
@@ -43,15 +43,19 @@ class InnerBlocks extends Component {
...
 		} = this.props;
 
-		const allowedBlocks = get( blockListSettings, [ 'allowedBlocks' ] );
-		if ( ! isShallowEqual( allowedBlocks, nextAllowedBlocks ) ) {
-			updateNestedSettings( {
-				allowedBlocks: nextAllowedBlocks,
-			} );
+		const newSettings = {
+			allowedBlocks,
+			templateLock: templateLock === undefined ? parentLock : templateLock,
```
Comment (aduth) [inline editor/components/inner-blocks/index.js]: Should it be the responsibility of someone rendering a `BlockList` to pass the `parentLock`, or should that component be smart enough to inherit?

If a parent block has locking but the child does not, which one wins?

**Review by aduth — APPROVED** [review]
Works well in testing

File: editor/components/inner-blocks/index.js (hunk 18 above)
Comment (jorgefilipecosta) [reply editor/components/inner-blocks/index.js]: When rendering a BlockList we don't pass the parentLock. We pass the templateLock. parentLock is just used for the case where templateLock is undefined and in that case we store parentLock in templateLock. Besides that case we never use parentLock again. 

> If a parent block has locking but the child does not, which one wins?

If the locking of an InnerBlocks area is undefined we use the locking of the parent. In all other cases, we use the templateLock passed even if the locking of the child is more permissive than the locking of the father.
Before we used always the less permissive locking of the tree https://github.com/WordPress/gutenberg/pull/6991#discussion_r191383168 but it was changed.

File: editor/store/selectors.js (hunk 17 above)
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: Unit tests were added.
